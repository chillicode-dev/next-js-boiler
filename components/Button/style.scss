.Button {
  display: inline-block;
  position: relative;
  padding: 10px 15px;
  background-color: $brandColor;
  color: white;
  outline: none;
  text-align: center;
  border: none;
  line-height: $lineHeight;
  cursor: pointer;
  transition: background-color $duration;

  &:hover {
    background-color: $brandColorDark;
  }
}

.link {
  @include link(white, white);
}

.loading {
  cursor: default;

  .content {
    visibility: hidden;
  }
}

.hasIcon {
  .content {
    display: flex;
    align-items: center;
  }
}

.loaderBox {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  @include fullsize;
}

.icon {
  margin-right: 5px;
}
