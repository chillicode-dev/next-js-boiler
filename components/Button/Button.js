// Vendor
import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
// Internals
import LoaderIcon from 'assets/icons/loader.svg';
import Link from 'components/Link';
import style from './style.scss';

const defaultIconSize = 25;

const Button = ({className, children, href, isLoading, icon, iconSize, ...props}) => {
  const classes = cn({
    [style.Button]: true,
    [style.loading]: isLoading,
    [style.link]: href,
    [style.hasIcon]: icon,
    [className]: className,
  });
  const TagName = href ? Link : 'button';
  const Icon = icon;
  const combinedProps = {
    className: classes,
    ...props,
  };
  if (href) {
    combinedProps.href = href;
  }

  return (
    <TagName {...combinedProps}>
      {isLoading && (
        <span className={style.loaderBox}>
          <LoaderIcon width={defaultIconSize} height={defaultIconSize} />
        </span>
      )}
      <span className={style.content}>
        {icon && <Icon width={iconSize} height={iconSize} className={style.icon} />}
        <span>{children}</span>
      </span>
    </TagName>
  );
};

Button.defaultProps = {
  href: '',
  isLoading: false,
  icon: '',
  iconSize: defaultIconSize,
  className: '',
};

Button.propTypes = {
  children: PropTypes.node.isRequired,
  href: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  isLoading: PropTypes.bool,
  icon: PropTypes.oneOfType([PropTypes.string, PropTypes.elementType]),
  iconSize: PropTypes.number,
  className: PropTypes.string,
};

export default Button;
