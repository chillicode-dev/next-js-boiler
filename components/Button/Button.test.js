import React from 'react';
import renderer from 'react-test-renderer';
import CheckIcon from 'assets/icons/check.svg';
import Button from './Button';

it('Renders Button correctly', () => {
  const elements = (
    <ul>
      <li>
        <Button>Click me</Button>
      </li>
      <li>
        <Button isLoading>Click me</Button>
      </li>
      <li>
        <Button href="//ya.ru" targetBlank>
          Link button
        </Button>
      </li>
      <li>
        <Button icon={CheckIcon} targetBlank>
          With icon
        </Button>
      </li>
      <li>
        <Button icon={CheckIcon} iconSize={40} targetBlank>
          With custom sized icon
        </Button>
      </li>
    </ul>
  );
  const tree = renderer.create(elements).toJSON();
  expect(tree).toMatchSnapshot();
});
