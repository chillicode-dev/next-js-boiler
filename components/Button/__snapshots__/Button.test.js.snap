// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Button correctly 1`] = `
<ul>
  <li>
    <button
      className="Button"
    >
      <span
        className="content"
      >
        
        <span>
          Click me
        </span>
      </span>
    </button>
  </li>
  <li>
    <button
      className="Button loading"
    >
      <span
        className="loaderBox"
      >
        <test-file-stub
          height={25}
          width={25}
        />
      </span>
      <span
        className="content"
      >
        
        <span>
          Click me
        </span>
      </span>
    </button>
  </li>
  <li>
    <a
      className="Button link"
      href="//ya.ru"
      rel="nofollow noopener"
      target="_blank"
    >
      <span
        className="content"
      >
        
        <span>
          Link button
        </span>
      </span>
    </a>
  </li>
  <li>
    <button
      className="Button hasIcon"
      targetBlank={true}
    >
      <span
        className="content"
      >
        <test-file-stub
          className="icon"
          height={25}
          width={25}
        />
        <span>
          With icon
        </span>
      </span>
    </button>
  </li>
  <li>
    <button
      className="Button hasIcon"
      targetBlank={true}
    >
      <span
        className="content"
      >
        <test-file-stub
          className="icon"
          height={40}
          width={40}
        />
        <span>
          With custom sized icon
        </span>
      </span>
    </button>
  </li>
</ul>
`;
